src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+            editor.InsertBefore(containingMethod, newMethodDeclaration);
+        }
+
+        /// <summary>
+        /// Gets the matches of the expression and replaces them with the identifier.
+        /// Special case for the original matching expression, if its parent is a LocalDeclarationStatement then it can
+        /// be removed.
+        /// Otherwise, it needs to have a rename annotation added to it.
+        /// </summary>
+        public static async Task UpdateExpressionInOriginalFunctionAsync(Document document,
+            TExpressionSyntax expression, SyntaxNode scope, string parameterName, SyntaxEditor editor,
+            bool allOccurrences, CancellationToken cancellationToken)
+        {
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var matches = await FindMatchesAsync(document, expression, scope, allOccurrences, cancellationToken).ConfigureAwait(false);
+            var replacement = (TIdentifierNameSyntax)generator.IdentifierName(parameterName);
+
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            foreach (var match in matches)
+            {
+                // Special case the removal of the originating expression to either remove the local declaration
+                // or to add a rename annotation.
+                if (!match.Equals(expression))
+                {
+                    editor.ReplaceNode(match, replacement);
+                }
+                else
+                {
+                    if (ShouldRemoveVariableDeclaratorContainingExpression(document, expression, out var varDeclName))
+                    {
+                        var localDeclaration = expression.FirstAncestorOrSelf<SyntaxNode>(node => syntaxFacts.IsLocalDeclarationStatement(node));
+                        Contract.ThrowIfNull(localDeclaration);
+                        editor.RemoveNode(localDeclaration);
+                    }
+                    else
+                    {
+                        // Getting the SyntaxToken and adding a rename annotation to the identifier.
+                        replacement = (TIdentifierNameSyntax)generator.IdentifierName(generator.Identifier(parameterName)
+                            .WithAdditionalAnnotations(RenameAnnotation.Create()));
+                        editor.ReplaceNode(match, replacement);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Goes through the parameters of the original method to get the location that the parameter
+        /// and argument should be introduced.
+        /// </summary>
+        private static int GetInsertionIndex(Compilation compilation, IMethodSymbol methodSymbol,
+            ISyntaxFactsService syntaxFacts, SyntaxNode methodDeclaration)
+        {
+            var parameterList = syntaxFacts.GetParameterList(methodDeclaration);
```

Comment:
do you ever use methodDeclaration *except* to get teh parameter list from it? if not, just pass the parameter list around.

Reply:
Also to figure out where the new methods (for the overload/trampoline) should be inserted in the editor.